pkg/controllers/pod_controller_test.go

```diff
@@ -0,0 +1,418 @@
+			leaderPod: leaderPodWrapper.AddAnnotation(jobset.ExclusiveKey, testTopologyKey).Obj(),
+			podList: &corev1.PodList{
+				Items: []corev1.Pod{
+					leaderPodWrapper.NodeSelector(nodeSelector).Obj(),
+					followerPodWrapper.NodeSelector(map[string]string{testTopologyKey: "topologyDomain"}).Obj(),
+				},
+			},
+			node: &corev1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   "test-node",
+					Labels: nodeSelector,
+				},
+			},
+			forceClientErr: true,
+			wantErr:        errors.Join(errors.New("example error")),
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			fc := makeFakeClient(interceptor.Funcs{
+				Get: func(ctx context.Context, client client.WithWatch,
+					key client.ObjectKey, obj client.Object, opts ...client.GetOption) error {
+					node, ok := obj.(*corev1.Node)
+					if !ok {
+						return nil
+					}
+					if tc.forceClientErr || node == nil {
+						return errors.Join(errors.New("example error"))
+					}
+					// Set returned node value to be the node defined in the test case.
+					*node = *tc.node
+					return nil
+				}})
+			r := &PodReconciler{
+				Client: fc.client,
+				Scheme: fc.scheme,
+				Record: fc.record,
+			}
+			gotMatched, gotErr := r.validatePodPlacements(context.Background(), &tc.leaderPod, tc.podList)
+			assert.Equal(t, tc.wantErr, gotErr)
+			assert.Equal(t, tc.wantMatched, gotMatched)
+		})
+	}
+}
+
+func TestDeleteFollowerPods(t *testing.T) {
+	var (
+		jobSetName       = "test-jobset"
+		ns               = "default"
+		leaderPodWrapper = makePod(&makePodArgs{
+			jobSetName:        jobSetName,
+			replicatedJobName: "replicated-job-1",
+			jobName:           "test-jobset-replicated-job-1-test-job-0",
+			podName:           "test-jobset-replicated-job-1-test-job-0-0",
+			ns:                ns,
+			nodeName:          "test-node",
+			jobIdx:            0}).
+			AddAnnotation(batchv1.JobCompletionIndexAnnotation, "0")
+		followerPodWrapper = makePod(&makePodArgs{
+			jobSetName:        jobSetName,
+			replicatedJobName: "replicated-job-1",
+			jobName:           "test-jobset-replicated-job-1-test-job-0",
+			podName:           "test-jobset-replicated-job-1-test-job-0-1",
+			ns:                ns,
+			jobIdx:            0})
+		wantPodWrapper = makePod(&makePodArgs{
+			jobSetName:        jobSetName,
+			replicatedJobName: "replicated-job-1",
+			jobName:           "test-jobset-replicated-job-1-test-job-0",
+			podName:           "test-jobset-replicated-job-1-test-job-0-1",
+			ns:                ns,
+			jobIdx:            0})
+	)
+	tests := []struct {
+		name            string
+		pods            []corev1.Pod
+		wantPodsDeleted []corev1.Pod
+		wantErr         error
+		forceClientErr  bool
+	}{
+		{
+			name: "delete follower pods",
+			pods: []corev1.Pod{
+				leaderPodWrapper.Obj(),
+				followerPodWrapper.Obj(),
+			},
+			wantPodsDeleted: []corev1.Pod{
+				wantPodWrapper.ResourceVersion("999").SetConditions([]corev1.PodCondition{{
```

Comment:
Can you remind me why we need to explicitly set the resourceVersion here?

Reply:
https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/client/fake/client.go#L285
I guess the reason for using a magic number here

Reply:
I see, honestly I think we should just count the number of pod deletion calls made via the Delete() interceptor func. Needing to dive into the weeds of another code base to understand the unit tests isn't worth it for the marginal improvement of the precision of our tests.

Sorry this came up rather late in the review process, I hope you don't mind making the change.

Reply:
No problem, it's also a learning experience for me, haha. I'll make the change later.

Reply:
fix. @danielvegamyhre